eth/db_upgrade.go

```diff
@@ -0,0 +1,300 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package eth implements the Ethereum protocol.
+package eth
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// upgradeSequentialKeys checks the chain database version and
+// starts a background process to make upgrades if necessary.
+// Returns a stop function that blocks until the process has
+// been safely stopped.
+func upgradeSequentialKeys(db ethdb.Database) (stopFn func()) {
+	var keyPtr []byte
+	var phase byte
+	storePtr := func() {
+		// store current ptr
+		db.Put([]byte("useSequentialKeys"), append([]byte{phase}, keyPtr...))
+	}
+	// read current ptr
+	data, _ := db.Get([]byte("useSequentialKeys"))
+	if len(data) > 0 {
+		phase = data[0]
+		keyPtr = data[1:]
+	}
+	if phase == 4 {
+		return nil // nothing to do
+	}
+
+	stopChn := make(chan struct{})
+	stopped := make(chan struct{})
+	go func() {
+		it := db.(*ethdb.LDBDatabase).NewIterator()
+		it.Seek(keyPtr)
+		if len(keyPtr) > 0 {
+			glog.V(logger.Info).Infof("Upgrading chain database to use sequential keys")
+			cnt, cc, cb, cr := 0, 0, 0, 0
+		loop:
+			for {
+				keyPtr := it.Key()
+				switch phase {
+				case 0:
+					phase++
+					it.Seek([]byte("block-num-"))
+				case 1:
+					if len(keyPtr) >= 10 && string(keyPtr[0:10]) == "block-num-" {
+						// if key is too long, it is probably a "block-" entry
+						// where the first 4 bytes of the hash happens to be "num-"
+						if len(keyPtr) < 20 {
+							cc++
+							if cc%100000 == 0 {
+								glog.V(logger.Info).Infof("converting %d canonical numbers...", cc)
+							}
+							number := big.NewInt(0).SetBytes(keyPtr[10:]).Uint64()
+							newKey := []byte("header-12345678-num")
+							binary.BigEndian.PutUint64(newKey[7:15], number)
+							db.Put(newKey, it.Value())
+							db.Delete(keyPtr)
+						}
+						it.Next()
+					} else {
+						glog.V(logger.Info).Infof("converted %d canonical numbers...", cc)
+						phase++
+						it.Release()
+						it = db.(*ethdb.LDBDatabase).NewIterator()
+						it.Seek([]byte("block-"))
+					}
+				case 2:
+					if len(keyPtr) >= 6 && string(keyPtr[0:6]) == "block-" {
+						if len(keyPtr) < 38 {
+							// invalid entry
+							db.Delete(keyPtr)
```

Comment:
Are you sure about this one? :P We store quit a few things in the database (e.g. trie nodes and txs are stored by hash, that's 32 bytes. This clause would delete all of them that happen to start with `block-`. :P